Mathlib/Tactic/TermCongr.lean

```diff
@@ -463,143 +479,218 @@ partial def mkCongrOf (depth : Nat) (mvarCounterSaved : Nat) (lhs rhs : Expr) :
-      | _ => panic! "unexpected hcongr argument kind"
-    let lhs := mkAppN fnRes.lhs lhsArgs
-    let rhs := mkAppN fnRes.rhs rhsArgs
-    if nontriv then
-      return CongrResult.mk' lhs rhs (mkAppN thm.proof args)
-    else
-      -- `lhs` and `rhs` *should* be defeq, but use `mkDefault` just to be safe.
-      CongrResult.mkDefault lhs rhs
-  | .lam .., .lam .. =>
-    trace[Elab.congr] "lam"
-    let resDom ← mkCongrOf (depth + 1) mvarCounterSaved lhs.bindingDomain! rhs.bindingDomain!
-    -- We do not yet support congruences in the binding domain for lambdas.
-    discard <| resDom.defeq
-    withLocalDecl lhs.bindingName! lhs.bindingInfo! resDom.lhs fun x => do
-      let lhsb := lhs.bindingBody!.instantiate1 x
-      let rhsb := rhs.bindingBody!.instantiate1 x
-      let resBody ← mkCongrOf (depth + 1) mvarCounterSaved lhsb rhsb
-      let lhs ← mkLambdaFVars #[x] resBody.lhs
-      let rhs ← mkLambdaFVars #[x] resBody.rhs
-      if resBody.isRfl then
-        return {lhs, rhs, pf? := none}
-      else
-        let pf ← mkLambdaFVars #[x] (← resBody.eq)
-        return CongrResult.mk' lhs rhs (← mkAppM ``funext #[pf])
-  | .forallE .., .forallE .. =>
-    trace[Elab.congr] "forallE"
-    let resDom ← mkCongrOf (depth + 1) mvarCounterSaved lhs.bindingDomain! rhs.bindingDomain!
-    if lhs.isArrow && rhs.isArrow then
-      let resBody ← mkCongrOf (depth + 1) mvarCounterSaved lhs.bindingBody! rhs.bindingBody!
-      let lhs := Expr.forallE lhs.bindingName! resDom.lhs resBody.lhs lhs.bindingInfo!
-      let rhs := Expr.forallE rhs.bindingName! resDom.rhs resBody.rhs rhs.bindingInfo!
-      if resDom.isRfl && resBody.isRfl then
-        return {lhs, rhs, pf? := none}
-      else
-        return CongrResult.mk' lhs rhs (← mkImpCongr (← resDom.eq) (← resBody.eq))
-    else
-      -- We do not yet support congruences in the binding domain for dependent pi types.
+  checkCache (lhs, rhs) fun _ => do
+    if let some res ← mkCongrOfCHole? mvarCounterSaved lhs rhs then
+      trace[Elab.congr] "hole processing succeeded"
+      return res
+    if lhs == rhs then
+      -- There should not be any cHoles, but to be safe let's remove them.
+      return { lhs := removeCHoles lhs, rhs := removeCHoles rhs, pf? := none }
+    if (hasCHole mvarCounterSaved lhs).isNone && (hasCHole mvarCounterSaved rhs).isNone then
+      -- It's safe to fastforward if the lhs and rhs are defeq and have no congruence holes.
+      -- This is more conservative than necessary since congruence holes might only be inside
+      -- proofs, and it is OK to ignore these.
+      if ← isDefEq lhs rhs then
+        return { lhs, rhs, pf? := none }
+    if ← (isProof lhs <||> isProof rhs) then
+      -- We don't want to look inside proofs at all.
+      return ← CongrResult.mkDefault lhs rhs
+    match lhs, rhs with
+    | .app .., .app .. =>
+      mkCongrOfApp depth mvarCounterSaved lhs rhs
+    | .lam .., .lam .. =>
+      trace[Elab.congr] "lam"
```

Comment:
I'd perhaps argue that `withTraceNode` here (and throughout) would be a bit more informative, as it would reflect the structure of the expr in the trace nesting.

Reply:
Ah, I turned on "hide whitespace changes" and now realize that this is not really part of the diff.